server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```
@@ -264,24 +270,90 @@ public void start()
+  public SegmentPublishResult announceHistoricalSegments(
+      final Set<DataSegment> segments,
+      final DataSourceMetadata startMetadata,
+      final DataSourceMetadata endMetadata
+  ) throws IOException
+  {
+    if (segments.isEmpty()) {
+      throw new IllegalArgumentException("segment set must not be empty");
+    }
+
+    final String dataSource = segments.iterator().next().getDataSource();
+    for (DataSegment segment : segments) {
+      if (!dataSource.equals(segment.getDataSource())) {
+        throw new IllegalArgumentException("segments must all be from the same dataSource");
+      }
+    }
+
+    if ((startMetadata == null && endMetadata != null) || (startMetadata != null && endMetadata == null)) {
+      throw new IllegalArgumentException("start/end metadata pair must be either null or non-null");
+    }
+
+    final AtomicBoolean txnFailure = new AtomicBoolean(false);
+
+    try {
+      return connector.retryTransaction(
+          new TransactionCallback<SegmentPublishResult>()
           {
-            final Set<DataSegment> inserted = Sets.newHashSet();
+            @Override
+            public SegmentPublishResult inTransaction(
+                final Handle handle,
+                final TransactionStatus transactionStatus
+            ) throws Exception
+            {
+              final Set<DataSegment> inserted = Sets.newHashSet();
+
+              if (startMetadata != null) {
+                final boolean success = updateDataSourceMetadataWithHandle(
+                    handle,
+                    dataSource,
+                    startMetadata,
+                    endMetadata
+                );
 
-            for (final DataSegment segment : segments) {
-              if (announceHistoricalSegment(handle, segment)) {
-                inserted.add(segment);
+                if (!success) {
+                  transactionStatus.setRollbackOnly();
+                  txnFailure.set(true);
+                  return null;
```

this return value is not used, as `transactionStatus.setRollbackOnly()` causes the txn to throw an exception rather than return a value. will make a note of this in a comment.